Add variable frame size support in Opus and Vorbis webm file

jyavenard opened the pull request:
#### 5296cfc10b4df1a6040139f4ca8d82f1343b30bf
<pre>
Add variable frame size support in Opus and Vorbis webm file
<a href="https://bugs.webkit.org/show_bug.cgi?id=261953">https://bugs.webkit.org/show_bug.cgi?id=261953</a>
rdar://99712672

Reviewed by Eric Carlson.

Adopt new CoreMedia support for variable size packets.
This is done by creating different magic cookies for Opus and Vorbis
stream.
And we then calculate the packet&apos;s frame size by parsing the first byte
of each packets according to the specs.

As this is only supported in iOS 17 and macOS Sonoma we keep the previous
way, which was just assuming all streams never changed.

Test added, on cocoa platforms, requires Sonoma or iOS 17.

* LayoutTests/media/content/opus_variable.webm: Added.
* LayoutTests/media/media-webm-opus-variable-length-expected.txt: Added.
* LayoutTests/media/media-webm-opus-variable-length.html: Added.
* LayoutTests/platform/ios-16/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::SharedBuffer::operator[] const):
* Source/WebCore/platform/SharedBuffer.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::PacketDurationParser::PacketDurationParser):
(WebCore::PacketDurationParser::framesInPacket):
(WebCore::PacketDurationParser::reset):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
(WebCore::SourceBufferParserWebM::setLogger):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::opusConfigToFrameDuration):
(WebCore::parseOpusTOCData):
(WebCore::parseOpusPrivateData):
(WebCore::cookieFromOpusCookieContents):
(WebCore::createOpusAudioInfo):
(WebCore::cookieFromVorbisCodecPrivate):
(WebCore::registerVorbisDecoderIfNeeded):
(WebCore::createVorbisAudioInfo):

Canonical link: <a href="https://commits.webkit.org/268588@main">https://commits.webkit.org/268588@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/211e62896ea88735604b1eac49ce7fdc26a555fe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20092 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20531 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21149 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21991 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18756 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20322 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23773 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20689 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20237 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20312 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20228 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17461 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22840 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17398 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18258 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24511 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/17473 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18475 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18434 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22501 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/19457 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19032 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16156 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/23496 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18228 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18094 "Passed tests") | [⏳ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Tests-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4823 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22572 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/24752 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18857 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5471 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

jkcalhoun commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -277,6 +282,179 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+        // It also corresponds to the delta in granule position of the packet within
+        // the same sequence of packets in an Ogg file, with the possible exception of
+        // the ultimate packet, which may be assigned a smaller delta for the purpose
+        // of trimming.
```
Too much information below this point. Let's delete all but the first paragraph of the comment.

jkcalhoun commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -1250,10 +1253,11 @@ webm::Status WebMParser::AudioTrackData::consumeFrameData(webm::Reader& reader,
+#if !HAVE(AUDIOFORMATPROPERTY_VARIABLEPACKET_SUPPORTED)
+    else if (codec() == CodecType::Opus) {
         // Opus technically allows the frame duration and frames-per-packet values to change from packet to packet.
         // CoreAudio doesn't support ASBD values like these to change on a per-packet basis, so throw an error when
```
Comment out of date. And as it turns out, it was never correct in the first place. It wasn't anything general about CoreAudio that imposed the constraint but the codec instead.

jkcalhoun commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -277,6 +282,179 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+            blockSize = m_vorbisModeInfo->mShortBlockSize;
+        else
+            blockSize = m_vorbisModeInfo->mLongBlockSize;
+        // The first vorbis packet decoded doesn't output audible content, and should be 0. But CoreMedia will choke on this.
```
0 isn't strictly correct even when the underlying media stack permits a 0 packet duration. What's really correct is: undetermined, because the true determination of the duration of any Vorbis packet requires one prior packet.

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
```diff
@@ -27,22 +27,52 @@
+    void reset();
+#if !HAVE(AUDIOFORMATPROPERTY_VARIABLEPACKET_SUPPORTED)
+    uint32_t framesPerPacket() const { return m_constantFramesPerPacket; }
+#endif
```
It looks like this is unused.

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -277,6 +283,129 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+    switch (m_audioFormatID) {
+#if ENABLE(VORBIS)
+    case kAudioFormatVorbis:
+        {
```
Nit: the brace should be on the previous line.

jyavenard replied:
webkit-style is so hopeless then. It whinged until I had it setup that way

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -277,6 +283,129 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+    switch (m_audioFormatID) {
+#if ENABLE(OPUS)
+    case kAudioFormatOpus:
+        {
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -277,6 +283,129 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+#endif
+#if ENABLE(VORBIS)
+    case kAudioFormatVorbis:
+        {
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -42,9 +45,12 @@
 #import <pal/cf/AudioToolboxSoftLink.h>
+#import <pal/cf/CoreMediaSoftLink.h>
 
+#if ENABLE(VORBIS)
```
Vorbis is always enabled on all platforms where we compile this file, so why do we need it in the file at all?

jyavenard replied:
I followed the pattern elsewhere, everything is conditional. We could probably remove it all, but that would be out of scope for this change.

jyavenard committed:
Add variable frame size support in Opus and Vorbis webm file
https://bugs.webkit.org/show_bug.cgi?id=261953
rdar://99712672

Reviewed by Eric Carlson.

Adopt new CoreMedia support for variable size packets.
This is done by creating different magic cookies for Opus and Vorbis
stream.
And we then calculate the packet's frame size by parsing the first byte
of each packets according to the specs.

As this is only supported in iOS 17 and macOS Sonoma we keep the previous
way, which was just assuming all streams never changed.

Test added, on cocoa platforms, requires Sonoma or iOS 17.

* LayoutTests/media/content/opus_variable.webm: Added.
* LayoutTests/media/media-webm-opus-variable-length-expected.txt: Added.
* LayoutTests/media/media-webm-opus-variable-length.html: Added.
* LayoutTests/platform/ios-16/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::SharedBuffer::operator[] const):
* Source/WebCore/platform/SharedBuffer.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::PacketDurationParser::PacketDurationParser):
(WebCore::PacketDurationParser::framesInPacket):
(WebCore::PacketDurationParser::reset):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
(WebCore::SourceBufferParserWebM::setLogger):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::opusConfigToFrameDuration):
(WebCore::parseOpusTOCData):
(WebCore::parseOpusPrivateData):
(WebCore::cookieFromOpusCookieContents):
(WebCore::createOpusAudioInfo):
(WebCore::cookieFromVorbisCodecPrivate):
(WebCore::registerVorbisDecoderIfNeeded):
(WebCore::createVorbisAudioInfo):

Canonical link: https://commits.webkit.org/268588@main
